### PR TITLE
Stop using atomic types in non-thread-safe contexts

### LINKED ISSUE #5812: Don't use Atomic* types where concurrency is not an issue
We use `AtomicReference` and `AtomicBoolean` etc. in many places to return values from visitors etc. Most of these cases have no use for atomicity, yet we still pay some price for the use of atomic data types. It is also somewhat confusing to see these types appear in contexts that should not have anything to do with concurrency.

### PR DESCRIPTION
**Pull request by lptr**
Fixes #5812 

- Replace `AtomicReference` with `MutableReference`
- Replace `AtomicBoolean` with `MutableBoolean`

Doing so makes it clear that there is no concurrency involved when reading the code, and the new `Mutable*` types are cheaper to execute.

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=13885198

File: subprojects/base-services/src/main/java/org/gradle/internal/MutableReference.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    public static <T> MutableReference<T> empty() {
+        return of(null);
+    }
+
+    public static <T> MutableReference<T> of(@Nullable T initialValue) {
+        return new MutableReference<T>(initialValue);
+    }
+
+    private MutableReference(@Nullable T initialValue) {
```
Comment (wolfs) [inline subprojects/base-services/src/main/java/org/gradle/internal/MutableReference.java]: Why do we use a static method to instantiate `MutableReference` and constructors for `MutableBoolean`? I think we should be consistent between the two. WDYT?

**Comment by lptr** [comment]
Another branch build: https://builds.gradle.org/viewQueued.html?itemId=13913179

File: subprojects/base-services/src/main/java/org/gradle/internal/MutableReference.java (hunk 1 above)
Comment (lptr) [reply subprojects/base-services/src/main/java/org/gradle/internal/MutableReference.java]: As we discussed, this is done for generics. A common pattern for the constructors is not a goal here.

File: subprojects/base-services/src/test/groovy/org/gradle/internal/MutableBooleanTest.groovy [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        then:
+        value.get()
+
+        when:
+        value.set(false)
+        then:
+        !value.get()
+    }
+
+    def "equals works"() {
```
Comment (wolfs) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/MutableBooleanTest.groovy]: I don't think `equals`, `hashCode` and `compare` add any value to `MutableBoolean`. Those things shouldn't be used in `Collection`s anyway, right? How about we remove those methods?

File: subprojects/base-services/src/test/groovy/org/gradle/internal/MutableReferenceTest.groovy [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        then:
+        ref.get() is object
+
+        when:
+        ref.set(null)
+        then:
+        ref.get() == null
+    }
+
+    def "equals works"() {
```
Comment (wolfs) [inline subprojects/base-services/src/test/groovy/org/gradle/internal/MutableReferenceTest.groovy]: Same as above. I would remove `equals` and `hashCode`.

**Comment by lptr** [comment]
Another branch build: https://builds.gradle.org/viewLog.html?buildId=13920891

**Pull request merged**